File: skills/cognitive-complexity/scripts/dart/lib/src/visitor.dart
import 'package:analyzer/dart/analysis/utilities.dart';
import 'package:analyzer/dart/ast/ast.dart';
import 'package:analyzer/dart/ast/token.dart';
import 'package:analyzer/dart/ast/visitor.dart';
import 'package:analyzer/source/line_info.dart';

import 'models.dart';

class DartComplexityAnalyzer {
  final int threshold;

  const DartComplexityAnalyzer({this.threshold = 15});

  FileComplexity analyzeSource(String sourceCode, {String filePath = '<stdin>'}) {
    final parseResult = parseString(content: sourceCode, throwIfDiagnostics: false);
    final unit = parseResult.unit;
    final lineInfo = parseResult.lineInfo;

    final collector = _FunctionCollector(lineInfo: lineInfo, threshold: threshold);
    unit.accept(collector);

    final functions = collector.functions;
    final totalComplexity = functions.fold<int>(0, (sum, f) => sum + f.complexity);
    final avgComplexity = functions.isNotEmpty ? totalComplexity / functions.length : 0.0;
    final highestComplexity = functions.fold<int>(0, (max, f) => f.complexity > max ? f.complexity : max);

    return FileComplexity(
      path: filePath,
      totalComplexity: totalComplexity,
      averageComplexity: avgComplexity,
      highestComplexity: highestComplexity,
      functions: functions,
    );
  }
}

class _FunctionCollector extends RecursiveAstVisitor<void> {
  final LineInfo lineInfo;
  final int threshold;
  final List<FunctionComplexity> functions = [];
  String? _currentClass;

  _FunctionCollector({required this.lineInfo, required this.threshold});

  @override
  void visitClassDeclaration(ClassDeclaration node) {
    final previousClass = _currentClass;
    _currentClass = node.name.lexeme;
    super.visitClassDeclaration(node);
    _currentClass = previousClass;
  }

  @override
  void visitMixinDeclaration(MixinDeclaration node) {
    final previousClass = _currentClass;
    _currentClass = node.name.lexeme;
    super.visitMixinDeclaration(node);
    _currentClass = previousClass;
  }

  @override
  void visitExtensionDeclaration(ExtensionDeclaration node) {
    final previousClass = _currentClass;
    _currentClass = node.name?.lexeme ?? '<anonymous_extension>';
    super.visitExtensionDeclaration(node);
    _currentClass = previousClass;
  }

  @override
  void visitFunctionDeclaration(FunctionDeclaration node) {
    final funcName = node.name.lexeme;
    final visitor = _FunctionComplexityVisitor(
      functionName: funcName,
      lineInfo: lineInfo,
      threshold: threshold,
    );
    node.functionExpression.body.accept(visitor);

    final startLine = lineInfo.getLocation(node.offset).lineNumber;
    final endLine = lineInfo.getLocation(node.end).lineNumber;

    functions.add(FunctionComplexity(
      name: funcName,
      className: _currentClass,
      lineNumber: startLine,
      endLineNumber: endLine,
      complexity: visitor.complexity,
      exceedsThreshold: visitor.complexity > threshold,
      breakdown: visitor.breakdown,
    ));

    super.visitFunctionDeclaration(node);
  }

  @override
  void visitMethodDeclaration(MethodDeclaration node) {
    final funcName = node.name.lexeme;
    final visitor = _FunctionComplexityVisitor(
      functionName: funcName,
      lineInfo: lineInfo,
      threshold: threshold,
    );
    node.body.accept(visitor);

    final startLine = lineInfo.getLocation(node.offset).lineNumber;
    final endLine = lineInfo.getLocation(node.end).lineNumber;

    functions.add(FunctionComplexity(
      name: funcName,
      className: _currentClass,
      lineNumber: startLine,
      endLineNumber: endLine,
      complexity: visitor.complexity,
      exceedsThreshold: visitor.complexity > threshold,
      breakdown: visitor.breakdown,
    ));

    super.visitMethodDeclaration(node);
  }

  @override
  void visitConstructorDeclaration(ConstructorDeclaration node) {
    final funcName = node.name?.lexeme ?? 'new';
    final visitor = _FunctionComplexityVisitor(
      functionName: funcName,
      lineInfo: lineInfo,
      threshold: threshold,
    );
    node.body.accept(visitor);

    final startLine = lineInfo.getLocation(node.offset).lineNumber;
    final endLine = lineInfo.getLocation(node.end).lineNumber;

    functions.add(FunctionComplexity(
      name: funcName,
      className: _currentClass,
      lineNumber: startLine,
      endLineNumber: endLine,
      complexity: visitor.complexity,
      exceedsThreshold: visitor.complexity > threshold,
      breakdown: visitor.breakdown,
    ));

    super.visitConstructorDeclaration(node);
  }
}

class _FunctionComplexityVisitor extends RecursiveAstVisitor<void> {
  final String functionName;
  final LineInfo lineInfo;
  final int threshold;

  int complexity = 0;
  int _currentNesting = 0;
  final List<ComplexityIncrement> breakdown = [];

  _FunctionComplexityVisitor({
    required this.functionName,
    required this.lineInfo,
    required this.threshold,
  });

  void _addIncrement(
    AstNode node,
    String nodeType,
    int baseIncrement,
    bool nestingPenalty,
    String reason,
  ) {
    final penalty = nestingPenalty ? _currentNesting : 0;
    final totalInc = baseIncrement + penalty;
    complexity += totalInc;

    final loc = lineInfo.getLocation(node.offset);
    final detail =
        '$reason (+$baseIncrement${penalty > 0 ? ' + nesting $penalty' : ''} = +$totalInc)';

    breakdown.add(ComplexityIncrement(
      line: loc.lineNumber,
      column: loc.columnNumber,
      nodeType: nodeType,
      increment: totalInc,
      nesting: _currentNesting,
      reason: detail,
    ));
  }

  void _processBoolOps(Expression expr, TokenType? parentOp) {
    Expression unwrapped = expr;
    while (unwrapped is ParenthesizedExpression) {
      unwrapped = unwrapped.expression;
    }

    if (unwrapped is BinaryExpression) {
      final op = unwrapped.operator.type;
      final isLogical = op == TokenType.AMPERSAND_AMPERSAND ||
          op == TokenType.BAR_BAR ||
          op == TokenType.QUESTION_QUESTION;

      if (isLogical) {
        final opName = op == TokenType.AMPERSAND_AMPERSAND
            ? '&&'
            : op == TokenType.BAR_BAR
                ? '||'
                : '??';

        if (parentOp == null) {
          _addIncrement(
              unwrapped, 'bool_op_sequence', 1, false, 'boolean operator sequence ($opName)');
        } else if (parentOp != op) {
          _addIncrement(
              unwrapped, 'bool_op_switch', 1, false, 'boolean operator switch to ($opName)');
        }

        _processBoolOps(unwrapped.leftOperand, op);
        _processBoolOps(unwrapped.rightOperand, op);
        return;
      }
    }
    unwrapped.accept(this);
  }

  @override
  void visitIfStatement(IfStatement node) {
    final parent = node.parent;
    final isElseIf = parent is IfStatement && parent.elseStatement == node;

    if (isElseIf) {
      _addIncrement(node, 'else_if', 1, false, 'else if branch');
    } else {
      _addIncrement(node, 'if', 1, true, 'if statement');
    }

    _processBoolOps(node.expression, null);

    _currentNesting++;
    node.thenStatement.accept(this);
    _currentNesting--;

    node.elseStatement?.accept(this);
  }

  @override
  void visitIfElement(IfElement node) {
    final parent = node.parent;
    final isElseIf = parent is IfElement && parent.elseElement == node;

    if (isElseIf) {
      _addIncrement(node, 'else_if_element', 1, false, 'else if element');
    } else {
      _addIncrement(node, 'if_element', 1, true, 'collection if element');
    }

    _processBoolOps(node.expression, null);

    _currentNesting++;
    node.thenElement.accept(this);
    _currentNesting--;

    node.elseElement?.accept(this);
  }

  @override
  void visitConditionalExpression(ConditionalExpression node) {
    _addIncrement(node, 'ternary', 1, true, 'ternary conditional expression');
    _processBoolOps(node.condition, null);

    _currentNesting++;
    node.thenExpression.accept(this);
    node.elseExpression.accept(this);
    _currentNesting--;
  }

  @override
  void visitSwitchStatement(SwitchStatement node) {
    _addIncrement(node, 'switch', 1, true, 'switch statement');
    node.expression.accept(this);

    _currentNesting++;
    for (final member in node.members) {
      member.accept(this);
    }
    _currentNesting--;
  }

  @override
  void visitSwitchExpression(SwitchExpression node) {
    _addIncrement(node, 'switch_expression', 1, true, 'switch expression');
    node.expression.accept(this);

    _currentNesting++;
    for (final member in node.cases) {
      member.accept(this);
    }
    _currentNesting--;
  }

  @override
  void visitForStatement(ForStatement node) {
    _addIncrement(node, 'for_loop', 1, true, 'for loop');
    _currentNesting++;
    node.body.accept(this);
    _currentNesting--;
  }

  @override
  void visitForElement(ForElement node) {
    _addIncrement(node, 'for_element', 1, true, 'collection for element');
    _currentNesting++;
    node.body.accept(this);
    _currentNesting--;
  }

  @override
  void visitWhileStatement(WhileStatement node) {
    _addIncrement(node, 'while_loop', 1, true, 'while loop');
    _currentNesting++;
    node.body.accept(this);
    _currentNesting--;
  }

  @override
  void visitDoStatement(DoStatement node) {
    _addIncrement(node, 'do_while_loop', 1, true, 'do-while loop');
    _currentNesting++;
    node.body.accept(this);
    _currentNesting--;
  }

  @override
  void visitCatchClause(CatchClause node) {
    _addIncrement(node, 'catch', 1, true, 'catch clause');
    _currentNesting++;
    node.body.accept(this);
    _currentNesting--;
  }

  @override
  void visitFunctionExpression(FunctionExpression node) {
    _currentNesting++;
    node.body.accept(this);
    _currentNesting--;
  }

  @override
  void visitMethodInvocation(MethodInvocation node) {
    if (node.methodName.name == functionName) {
      _addIncrement(node, 'recursion', 1, false, 'direct recursion call');
    }
    super.visitMethodInvocation(node);
  }

  @override
  void visitFunctionExpressionInvocation(FunctionExpressionInvocation node) {
    final function = node.function;
    if (function is SimpleIdentifier && function.name == functionName) {
      _addIncrement(node, 'recursion', 1, false, 'direct recursion call');
    }
    super.visitFunctionExpressionInvocation(node);
  }
}

File: skills/cognitive-complexity/scripts/typescript/src/visitor.ts
import * as ts from "typescript";
import {
  ComplexityIncrement,
  ComplexityReport,
  FileComplexity,
  FunctionComplexity,
} from "./types";

export class TypeScriptComplexityAnalyzer {
  constructor(private threshold: number = 15) {}

  public analyzeSource(sourceCode: string, filePath: string = "<stdin>"): FileComplexity {
    const sourceFile = ts.createSourceFile(
      filePath,
      sourceCode,
      ts.ScriptTarget.Latest,
      true,
      filePath.endsWith(".tsx") || filePath.endsWith(".jsx") ? ts.ScriptKind.TSX : ts.ScriptKind.TS
    );

    const functions: FunctionComplexity[] = [];
    this.collectFunctions(sourceFile, sourceFile, undefined, functions);

    const totalComplexity = functions.reduce((acc, f) => acc + f.complexity, 0);
    const avgComplexity = functions.length ? totalComplexity / functions.length : 0;
    const highestComplexity = functions.reduce((max, f) => Math.max(max, f.complexity), 0);

    return {
      path: filePath,
      total_complexity: totalComplexity,
      average_complexity: Math.round(avgComplexity * 100) / 100,
      highest_complexity: highestComplexity,
      functions,
    };
  }

  private collectFunctions(
    node: ts.Node,
    sourceFile: ts.SourceFile,
    className: string | undefined,
    functions: FunctionComplexity[]
  ): void {
    if (ts.isClassDeclaration(node) || ts.isClassExpression(node) || ts.isInterfaceDeclaration(node)) {
      const currentClassName = node.name?.text || "AnonymousClass";
      const fullClassName = className ? `${className}.${currentClassName}` : currentClassName;
      ts.forEachChild(node, (child) => this.collectFunctions(child, sourceFile, fullClassName, functions));
      return;
    }

    const funcInfo = this.extractFunctionInfo(node, sourceFile, className);
    if (funcInfo) {
      functions.push(this.analyzeFunctionNode(funcInfo.node, sourceFile, funcInfo.name, className));
    }

    ts.forEachChild(node, (child) => this.collectFunctions(child, sourceFile, className, functions));
  }

  private extractFunctionInfo(
    node: ts.Node,
    sourceFile: ts.SourceFile,
    className?: string
  ): { node: ts.Node; name: string } | null {
    if (ts.isFunctionDeclaration(node) && node.name) {
      return { node, name: node.name.text };
    }
    if (ts.isMethodDeclaration(node) && node.name) {
      return { node, name: node.name.getText(sourceFile) };
    }
    if (ts.isConstructorDeclaration(node)) {
      return { node, name: "constructor" };
    }
    if (ts.isGetAccessorDeclaration(node) || ts.isSetAccessorDeclaration(node)) {
      return { node, name: node.name.getText(sourceFile) };
    }
    if (ts.isVariableDeclaration(node) && node.initializer && (ts.isArrowFunction(node.initializer) || ts.isFunctionExpression(node.initializer))) {
      return { node: node.initializer, name: node.name.getText(sourceFile) };
    }
    return null;
  }

  public analyzeFunctionNode(
    functionNode: ts.Node,
    sourceFile: ts.SourceFile,
    functionName: string,
    className?: string
  ): FunctionComplexity {
    let complexity = 0;
    const breakdown: ComplexityIncrement[] = [];

    const addIncrement = (node: ts.Node, type: string, base: number, nesting: number, reason: string) => {
      const total = base + nesting;
      complexity += total;
      const { line, character } = sourceFile.getLineAndCharacterOfPosition(node.getStart(sourceFile));
      breakdown.push({
        line: line + 1,
        column: character + 1,
        type,
        increment: total,
        nesting,
        reason: `${reason} (+${base}${nesting > 0 ? ` + nesting ${nesting}` : ""} = +${total})`,
      });
    };

    const visitor = new FunctionAstVisitor(sourceFile, functionName, addIncrement);
    const body = (functionNode as any).body || functionNode;
    visitor.visit(body, 0);

    const { line: startLine } = sourceFile.getLineAndCharacterOfPosition(functionNode.getStart(sourceFile));
    const { line: endLine } = sourceFile.getLineAndCharacterOfPosition(functionNode.getEnd());

    return {
      name: functionName,
      class_name: className || null,
      line_number: startLine + 1,
      end_line_number: endLine + 1,
      complexity,
      exceeds_threshold: complexity > this.threshold,
      breakdown,
    };
  }
}

class FunctionAstVisitor {
  constructor(
    private sourceFile: ts.SourceFile,
    private functionName: string,
    private addIncrement: (node: ts.Node, type: string, base: number, nesting: number, reason: string) => void
  ) {}

  public visit(node: ts.Node, nesting: number): void {
    if (ts.isIfStatement(node)) {
      this.handleIf(node, nesting);
    } else if (ts.isConditionalExpression(node)) {
      this.handleConditional(node, nesting);
    } else if (ts.isForStatement(node) || ts.isForInStatement(node) || ts.isForOfStatement(node) || ts.isWhileStatement(node) || ts.isDoStatement(node)) {
      this.handleLoop(node, nesting);
    } else if (ts.isSwitchStatement(node)) {
      this.handleSwitch(node, nesting);
    } else if (ts.isCatchClause(node)) {
      this.handleCatch(node, nesting);
    } else if (ts.isBreakStatement(node) || ts.isContinueStatement(node)) {
      this.handleJump(node);
    } else if (ts.isCallExpression(node)) {
      this.handleCall(node, nesting);
    } else if (ts.isArrowFunction(node) || ts.isFunctionExpression(node)) {
      this.handleClosure(node, nesting);
    } else {
      ts.forEachChild(node, (child) => this.visit(child, nesting));
    }
  }

  private handleIf(node: ts.IfStatement, nesting: number): void {
    const isElseIf = node.parent && ts.isIfStatement(node.parent) && node.parent.elseStatement === node;
    if (isElseIf) {
      this.addIncrement(node, "else_if", 1, 0, "else if statement");
    } else {
      this.addIncrement(node, "if", 1, nesting, "if statement");
    }

    this.processBoolOps(node.expression, undefined);
    this.visit(node.thenStatement, isElseIf ? nesting : nesting + 1);

    if (node.elseStatement) {
      if (ts.isIfStatement(node.elseStatement)) {
        this.visit(node.elseStatement, nesting);
      } else {
        this.visit(node.elseStatement, nesting + 1);
      }
    }
  }

  private handleConditional(node: ts.ConditionalExpression, nesting: number): void {
    this.addIncrement(node, "ternary", 1, nesting, "ternary conditional expression");
    this.processBoolOps(node.condition, undefined);
    this.visit(node.whenTrue, nesting + 1);
    this.visit(node.whenFalse, nesting + 1);
  }

  private handleLoop(node: ts.Node, nesting: number): void {
    this.addIncrement(node, "loop", 1, nesting, "loop statement");
    ts.forEachChild(node, (child) => {
      if (child === (node as any).statement) {
        this.visit(child, nesting + 1);
      } else {
        this.visit(child, nesting);
      }
    });
  }

  private handleSwitch(node: ts.SwitchStatement, nesting: number): void {
    this.addIncrement(node, "switch", 1, nesting, "switch statement");
    this.visit(node.expression, nesting);
    this.visit(node.caseBlock, nesting + 1);
  }

  private handleCatch(node: ts.CatchClause, nesting: number): void {
    this.addIncrement(node, "catch", 1, nesting, "catch block");
    this.visit(node.block, nesting + 1);
  }

  private handleJump(node: ts.BreakStatement | ts.ContinueStatement): void {
    if (node.label) {
      this.addIncrement(node, "labeled_jump", 1, 0, "labeled break/continue");
    }
  }

  private handleCall(node: ts.CallExpression, nesting: number): void {
    if (ts.isIdentifier(node.expression) && node.expression.text === this.functionName) {
      this.addIncrement(node, "recursion", 1, 0, "direct recursion call");
    }
    ts.forEachChild(node, (child) => this.visit(child, nesting));
  }

  private handleClosure(node: ts.Node, nesting: number): void {
    const body = (node as any).body;
    if (body) {
      this.visit(body, nesting + 1);
    }
  }

  private unwrapParentheses(node: ts.Node): ts.Node {
    let current = node;
    while (ts.isParenthesizedExpression(current)) {
      current = current.expression;
    }
    return current;
  }

  private processBoolOps(node: ts.Node, currentOp: ts.SyntaxKind | undefined): void {
    const expr = this.unwrapParentheses(node);
    if (!ts.isBinaryExpression(expr)) {
      this.visit(expr, 0);
      return;
    }

    const op = expr.operatorToken.kind;
    const isLogical = op === ts.SyntaxKind.AmpersandAmpersandToken || op === ts.SyntaxKind.BarBarToken || op === ts.SyntaxKind.QuestionQuestionToken;

    if (!isLogical) {
      this.visit(expr.left, 0);
      this.visit(expr.right, 0);
      return;
    }

    const opName = op === ts.SyntaxKind.AmpersandAmpersandToken ? "&&" : op === ts.SyntaxKind.BarBarToken ? "||" : "??";
    if (currentOp === undefined) {
      this.addIncrement(expr, "bool_op_sequence", 1, 0, `boolean operator sequence (${opName})`);
    } else if (currentOp !== op) {
      this.addIncrement(expr, "bool_op_switch", 1, 0, `boolean operator switch to (${opName})`);
    }

    this.processBoolOps(expr.left, op);
    this.processBoolOps(expr.right, op);
  }
}
